// class/beings/Aguila.js
exports.Aguila = function( params ){

	// Props del hijo (con mas valor que las del Padre)
	var defaults = {
		"sprite":{
			"images":"../public/sprites/testsprites.png",
			"hasToCalculatePosition":true,
			"width": 14,
			"height": 14,
			"rotation": 0,
			"offsetHeight":7,
			"animation": {
				"up": {
	                offset:[0, 0],
	                size:[32*3, 32],
	                frameWidth: 32,
	                frameHeight: 32,
	                framesCount: 3,
	                speed: 200,           //full animation time ms.
	                rotationAroundX: 360,   //degrees
	                spriteRotation: 0     //degrees
	            },
	            "right": {
	                offset:[0, 96],
	                size:[32*3, 32],
	                frameWidth: 32,
	                frameHeight: 32,
	                framesCount: 3,
	                speed: 200,           //full animation time ms.
	                rotationAroundX: 360,   //degrees
	                spriteRotation: 0     //degrees
	            },
	            "down": {
	                offset:[0, 32],
	                size:[32*3, 32],
	                frameWidth: 32,
	                frameHeight: 32,
	                framesCount: 3,
	                speed: 200,           //full animation time ms.
	                rotationAroundX: 360,   //degrees
	                spriteRotation: 0     //degrees
	            },
	            "left": {
	                offset:[0, 64],
	                size:[32*3, 32],
	                frameWidth: 32,
	                frameHeight: 32,
	                framesCount: 3,
	                speed: 200,           //full animation time ms.
	                rotationAroundX: 360,   //degrees
	                spriteRotation: 0     //degrees
	            },
	        }
		},
		/*Nombre del elfo*/
		"name":"Aguila",
		"specie":"Aguila",
		"stats" : {
			"life" : { "min" : 15 , "max" : 15 },

			"damage": { "min" : 2 , "max" : 3 },

			"shields": { "min" : 0 , "max" : 0 },

			"accurancy" : 100e3,

			"agility" : 10e3,

			"points": {"min":210 }
		}
	}

	// Hereda del padre y le agrega defaults.
	extend( this , Being , defaults );

	// Sobreescribe las propiedades con los valores pasados en params cuando se instancia.
	merge( this , params , this );

	this.id = this.generateId();

	// Se agrega en el mapa
	world.getTile( this.position ).append({ 'being' : this });
	this.createControls();

}
