// class/Server.js
//------------------------------------------------------------------
// Server
//------------------------------------------------------------------

// SERVER
exports.Server = function(){

	// Fecha de inicio del server.
	this.startDate = (new Date()).getTime();

	// Lista de datos que le pasa a los users
	this.queueList = [];
	/*
		Being.move()
		Being.isAffected();
	*/

	this.initClient = function( params ){
		var data = { "world" : world };
		params.socket.emit( 'init_client' , data );
	}

	// Agrega a la cola de informacion que manda al user.
	this.queue = function( params ){
		this.queueList.push( params );
	}


	this.queueCommit = function(){
		test('Update World');
		for( var item in this.queueList ){
			// Envia datos a todos los conectados.
			for( var u in USERS.users ){
				var data = this.queueList[item];
				USERS.users[u].socket.emit( 'world_update' , data );
			}
		}
		this.queueList = [];
	}


	this.refresh = setInterval( this.queueCommit.bind(this) , 500);


}

console.log('Clase Server -> OK');
