// class/Server.js
//------------------------------------------------------------------
// Server
//------------------------------------------------------------------

// SERVER
exports.Server = function(){

	// Fecha de inicio del server.
	this.startDate = (new Date()).getTime();

	// Lista de datos que le pasa a los users
	this.queueList = {};
	/*
		Being.move()
		Being.isAffected();
	*/

	this.initClient = function( params ){
		setInterval(function(){

			var data = { "world" : world };
			params.socket.emit( 'init_client' , data );
			data={};
		},500)
	}

	// Agrega a la cola de informacion que manda al user.
	this.queue = function( params ){
		if( this.queueList[ params.user ] == undefined ) this.queueList[ params.user ] = {};
		this.queueList[ params.user ].controls = params.controls;
		this.queueList[ params.user ][ params.data.all[0].three ] = params.data.all[0].value;
		test("listado");
		console.log(this.queueList);
		test('-------')
	}


	this.queueCommit = function(){

		var user;
		for( var item in this.queueList ){

			// Verifica si es pc o npc.
			if( this.queueList[item].controls == 'pc' ){
				user = USERS.getByUsername({ "username" : item });
			}else{
				user = item;
			}

			// Si todavia esta logueado el autor de la accion..
			if( user ){
				test('Update World');
				// Envia datos a todos los conectados.
				for( var u in USERS.users ){
					var data = { "author" : item , "data" : this.queueList[item] }
					USERS.users[u].socket.emit( 'world_update' , data );
				}
			}
		}
		this.queueList = {};
	}


	this.refresh = setInterval( this.queueCommit.bind(this) , 500);


}

console.log('Clase Server -> OK');
