// class/Users.js
//------------------------------------------------------------------
// USERS
//------------------------------------------------------------------

exports.Users = function(){
	
	this.users = {}

	// USERS.connect({ "user" : User })
	// Pasar ademas el objeto socket de la conexion a User.
	this.connect = function( params ){
		return this.users[ params.user.username ] = params.user;
	}

	// USERS.disconnect({ "user" : User })
	this.disconnect = function( params ){
		// Borro el being del mapa.
		var being = world.findBeing({ "userId" : params.user.username });
		being.remove();
		// Borro usuario de conectados.
		return ( delete this.users[ params.user.username ] );
	}

	// USERS.disconnectAll()
	this.disconnectAll = function(){
		return this.users = {};
	}

	// USERS.get({ "user" : User })
	this.get = function ( params ){
		return this.users[ params.user.username ];
	}

	// USERS.get({ "user" : User })
	// Encuentra user por username
	this.getByUsername = function ( params ){
		return this.users[ params.username ];
	}

	this.getSocket = function( params ){
		return this.users[ params.userId ].socket;
	}

	// USERS.ping({ "user" : User })
	// Checkea conexion con un usuario inactivo. Si no responde lo desconecta
	this.ping = function( params ){
	}

	// USERS.isConnected({ "user" : User })
	// Checkea conexion con un usuario.
	this.isConnected = function( params ){
	}

	// USERS.message({ "user" : User , "message" : "world save" });
	this.message = function( params ){
	}

	// USERS.messageAll({ "message" : "world save" });
	this.messageAll = function( params ){
	}


}

console.log('Clase USERS -> OK');
